README.md
# Databox
The Databox platform is an open-source personal networked device, augmented by cloud-hosted services, that collates, curates, and mediates access to an individual’s personal data by verified and audited third-party applications and services. The Databox will form the heart of an individual’s personal data processing ecosystem, providing a platform for managing secure access to data and enabling authorised third parties to provide the owner with authenticated services, including services that may be accessed while roaming outside the home environment. Databox project is led by Dr Hamed Haddadi (Imperial College) in collaboration with Dr Richard Mortier (University of Cambridge) and Professors Derek McAuley, Tom Rodden, Chris Greenhalgh, and Andy Crabtree (University of Nottingham) and funded by EPSRC. See http://www.databoxproject.uk/ for more information.

## Getting Started

These instructions will get a copy of the Databox up and running on your local machine. For development and testing purposes, see Development section below.

### Prerequisites

1) Requires Docker. Read [here](https://docs.docker.com/engine/installation/) for docker installation.

> Note: currently supported platforms are Linux and MacOS. Running on other platforms is possible using a virtual machine running Linux with bridge mode networking. Also note that more than one CPU core must be allocated to the VM.

### Get started
Make sure Docker is installed and running before starting Databox.  Run the following to get your databox up and
running.

```
docker run --rm -v /var/run/docker.sock:/var/run/docker.sock --network host -t databoxsystems/databox:0.5.1 /databox start -sslHostName $(hostname)
```

> Note: arm64v8 Platforms must be running a 64 bit version of linux (Alpine 3.8 aarch64)[https://alpinelinux.org/downloads/] or (HypriotOS/arm64)[https://github.com/DieterReuter/image-builder-rpi64/releases]

The above starts Databox using pre-build images published on [Docker hub](<https://hub.docker.com/r/databoxsystems>) and runs Databox on your local machine.

Once it's started, point a web browser at <http://127.0.0.1> and follow the instructions to configure your HTTPS certificates to access Databox UI securely (using a web browser <https://127.0.0.1>, or the iOS and Android app).

> Note: Using the databox iOS and Android apps with MacOS may require you to modify your firewall to enable external access to port 80 and 443.

To stop databox and clean up,
```
docker run --rm -v /var/run/docker.sock:/var/run/docker.sock -t databoxsystems/databox:0.5.1 /databox stop
```

# Development

## Get Started with the Graphical SDK

The graphical SDK will allow you to quickly build and test simple databox apps. To start the SDK run:
```
docker run --rm -v /var/run/docker.sock:/var/run/docker.sock --network host -t databoxsystems/databox:0.5.1 /databox sdk -start
```
The SDK web UI is available at http://127.0.0.1:8086

To stop the SDK run:
```
docker run --rm -v /var/run/docker.sock:/var/run/docker.sock --network host -t databoxsystems/databox:0.5.1 /databox sdk -stop
```

## Developing apps and drivers without the SDK

It is possible to develop Databox apps and driver without the SDK. Currently, [Python](https://github.com/me-box/lib-python-databox), [Golang](https://github.com/me-box/lib-go-databox) and [NodeJs](https://github.com/me-box/lib-node-databox) all have support libraries. Building outside the SDK allows you to make smaller more efficient containers and use more third-party libraries.

To get started all you need is a Dockerfile and a databox-manifest.json examples can be found in the libraries '/samples' directories. To make your app available to install locally on your databox you will need to upload the app-store driver and use `docker build -t [your-app-name] .`. Once the manifest is uploaded and the image has built then you should be up to install the app on your local Databox.

A good place to get started is the [databox quickstart repo](https://github.com/me-box/databox-quickstart/) which has all you need to develop apps and drivers and a small tutorial.

>>Images must be post fixed with -amd64 or -arm64v8 respectively.
>>The image must have the version tag that matches your running version of databox :0.5.1 or :latest for example.

If you would like to modify one of the currently available actual drivers you can do so by doing the following:
```
./databox-install-component driver-os-monitor
```

This will download and build the code on your machine and upload the Databox manifest to your local app store. You can also use this with your repositories and forks using:
```
./databox-install-component [GITHUB_USERNAME]/[GITHUB_REPONAME]
```

## Developing core components

To develop on the platform and core components the databox start command allows you to replace the databoxsystems core images with your owen. For example to replace the arbiter.

```
docker build databoxdev/arbiter .                                     # build your updated arbiter image
make start OPTS=--release 0.5.1 --arbiter databoxdev/arbiter      # start databox using the new code
```

# Databox Components

Databox has a number of platform components, divided into two parts:  Core and Other components.  Core components are required for Databox function.  Other components of things like apps and drivers to demonstrate Databoxes functionality.

## Core

* [Databox-container-manager](https://github.com/me-box/core-container-manager) Container manager controls build, installation and running functions of the other databox components.
* [databox-arbiter](https://github.com/me-box/core-arbiter) Arbiter manages the flow of data by minting tokens and controlling store discovery.
* [databox-export-service](https://github.com/me-box/core-export-service) This service controls the data to be exported to external URLs.
* [core-store](https://github.com/me-box/core-store)  This is a data store used by apps and drivers to store and retrieve JSON data or JPEG images.
* [core-ui](https://github.com/me-box/core-ui)  This is the databox default user interface.
* [driver-app-store](https://github.com/me-box/driver-app-store) This is a driver for retrieving manifests and making them available to your databox.

## Other

### Drivers
* [driver-sensingkit](https://github.com/me-box/driver-sensingkit) This driver provides SensingKit mobile sensor data.
* [driver-google-takeout](https://github.com/me-box/driver-google-takeout) This driver supports bulk import of google takeout data.
* [driver-phillips-hue](https://github.com/me-box/driver-phillips-hue) This drivers allows connection to Phillips Hue Platform.
* [driver-os-monitor](https://github.com/me-box/driver-os-monitor) This driver monitors the databox hardware by fetching Memory consumption and CPU load.
* [driver-twitter](https://github.com/me-box/driver-twitter) This driver streams data from a twitter account into a datastore.
* [driver-tplink-smart-plug](https://github.com/me-box/driver-tplink-smart-plug) This driver collects data from TP-Link smart plugs.
### Apps
* [app-light-graph](https://github.com/me-box/app-light-graph) An app that plots mobile phone light sensor data.
* [app-twitter-sentiment](https://github.com/me-box/app-twitter-sentiment) An app that used data from driver-twitter to calculate tweet sentiment.
* [app-os-monitor](https://github.com/me-box/app-os-monitor) An app to plot the output of the data generated by [driver-os-monitor](https://github.com/me-box/driver-os-monitor).

## Libraries for writing drivers and apps
For writing a new driver or app for Databox, one needs [Databox APIs](./documents/api_specification.md). To make app/driver development easy, we have wrapped Databox APIs in [nodejs](https://nodejs.org/en/), [python](https://docs.python.org/3.4/library/index.html) and [go](https://golang.org/). Using any of these libraries, a developer can build their databox app/driver.
* [lib-node-databox](https://github.com/me-box/node-databox): Databox Nodejs API library for building databox apps and drivers.
* [lib-python-databox](https://github.com/me-box/lib-python-databox): Databox Python API library for building databox apps and drivers.
* [lib-go-databox](https://github.com/me-box/lib-go-databox): Databox Go API library for building databox apps and drivers.
#### API and System specifications
Databox System Design document can be find [here](./documents/system_overview.md) and general API specifications are [here](./documents/api_specification.md).

## Setting up a full development clone of databox

>> Multi arch builds only work on Docker for Mac experimental
>> enable docker cli experimental features "experimental": "enabled" ~/.docker/config.json
```
    make all ARCH=amd64 DEFAULT_REG=[your docker hub reg tag]
```
Or for amd64v8 platforms
'''
    make all ARCH=amd64v8 DEFAULT_REG=[your docker hub reg tag]
'''

## Running the tests

```
make test

```
For more details, have a look [here](./TESTING.md).

## Contributing

The databox project welcomes contributions via pull requests see [CONTRIBUTING.md](./CONTRIBUTING.md) for more information. A good start is to look at the current [issues](https://github.com/me-box/databox/issues) and [forking](https://github.com/me-box/databox#fork-destination-box) the databox repo and fixing bugs/issues and submitting a pull request. Read more on Fork and Pull [here](https://help.github.com/articles/fork-a-repo/).

## Versioning

This documentation is up-to-date till this [commit](https://github.com/me-box/databox/tree/45bb7b2f47f595d12a952d9902ffd99061dae6eb). The master branches on all components point to the current release and are tagged in git using [semver](http://semver.org/).

## Authors

The list of [contributors](https://github.com/me-box/databox/contributors) who participated in this project.

## License
MIT Licence, See [here](./LICENSE).

## Contributing

The Databox project welcomes contributions via pull requests see [CONTRIBUTING.md](./CONTRIBUTING.md) for more information.

## Development of databox was supported by the following funding

```
EP/N028260/1, Databox: Privacy-Aware Infrastructure for Managing Personal Data

EP/N028260/2, Databox: Privacy-Aware Infrastructure for Managing Personal Data

EP/N014243/1, Future Everyday Interaction with the Autonomous Internet of Things

EP/M001636/1, Privacy-by-Design: Building Accountability into the Internet of Things (IoTDatabox)

EP/M02315X/1, From Human Data to Personal Experience

```


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [0.5.1]

This version contains a number of bug fixes and some new features including:

  -  fixed bugs #291, #289, #285 #284 #277
  -  Version manifests and allow addition of manifest repos to the app-store driver
  -  Manifests for apps and drivers can now provide a full docker image name and tag
  -  Changes to the manifest format to enable core-ui to install dependencies at install time
  -  Fixed nodejs builds
  -  Containers running as root (all apps and drivers now run as an un-privileged user)
  -  first version of the quickstart guide https://github.com/me-box/databox-quickstart
  -  make databox recover from host and docker demon restarts
  -  App and Driver crashes are detected and restarted cleanly
  -  removed the need for the CM and SDK to bind mount directories on the host
  -  updates to the arbiter, fixes export token support
  -  The export service has ben re-enabled
  -  Arm64v8 support has been enabled and arm images built. Databox now runs on a Pi3b+ (testing and performance improvements needed)


## [0.5.0]

This version contains a rewritten core-arbiter, core-container-manager and a new set of build tools. There is now no javascript in the code of databox. Most of the core databox components now communicate over the zest protocol, the language-specific libraries have been updated to reflect this change.

The databox user interface has been moved for the core-container-manager into its own component core-ui. This uses a new experimental store based API to access data and API endpoints within the container-manager. This enhances security and enables audit logging by default. It also has the benefit that new user interfaces can be developed in the same way as databox apps.

The platform-app server has also been removed in favour of a databox driver that read manifest from a git hub repository (driver-app-store).

With these changes, the core of databox is much more stable and should be easier or extend and develop on in the future.

Have fun, and as always expect bugs and dragons. Please report issues on the main me-box/databox repository.

The Databox Team

## [0.2.0]
### Changed
    * renamed all the repos see #52
    * Re-enable docker hub builds
    * removed need for remote app-store (for now)
    * removed syslog loging will be done by stores
    * new UI and Android app

## [0.1.2]
### Changed
    * Moved to docker swarm mode #42
    * Removed need for local registry for development
    * pass HTTPS certs and arbiter keys as docker secrets #3
    * moved the container manager into it own repo me-box/databox-cm
    * Removed some modules and implemented mod clean to reduce container manager image size
    * Docker Network API issue - when creating network first time. Fixed #29
    * Integrated the graphical SDK

## [0.1.1] 2017-04-05
### Changed
    * Moved databox container manager to node:alpine to save bandwidth. #14
    * Removed the need to run update updateLocalRegistry.sh #13
    * code cleanup in container-manager.js
    * Update to local registry Env Vars
    * reduced size of databox-export service

## [0.1.0] 2017-03-24

    * The fist open source release. Provides basic versions of all core components and a set of APIs on which to build databox Apps and drivers

main.go
package main

import (
	"context"
	"flag"
	"fmt"
	"io"
	"io/ioutil"
	"net"
	"net/http"
	"os"
	"runtime"
	"strings"
	"syscall"
	"time"

	"encoding/json"

	"github.com/docker/docker/api/types"
	"github.com/docker/docker/api/types/filters"
	"github.com/docker/docker/api/types/mount"
	"github.com/docker/docker/api/types/swarm"
	"github.com/docker/docker/client"
	libDatabox "github.com/me-box/lib-go-databox"
)

var path string
var dockerCli *client.Client

const certsBasePath = "./certs"
const CURRENT_RELEASE = "0.5.1"
const DEFAULT_REGISTRY = "databoxsystems"

func main() {

	DOCKER_API_VERSION := flag.String("API", "1.37", "Docker API version ")

	startCmd := flag.NewFlagSet("start", flag.ExitOnError)
	startCmdIP := startCmd.String("swarm-ip", "127.0.0.1", "The IP on the host to advertise the swarm.")
	startCmdRelease := startCmd.String("release", CURRENT_RELEASE, "Databox version to start, can uses tagged versions or latest")
	startCmdRegistryHosts := startCmd.String("registryHost", "docker.io", "Override the default registry host, server where images are pulled form")
	startCmdRegistry := startCmd.String("registry", DEFAULT_REGISTRY, "Override the default registry path, where images are pulled form")
	startCmdPassword := startCmd.String("password", "", "Override the password if you dont want an auto generated one. Mainly for testing")
	appStore := startCmd.String("manifestStore", "https://github.com/me-box/databox-manifest-store", "Override the default manifest store where manifests are loaded form (must be a repo on github)")
	cmImage := startCmd.String("cm", "databoxsystems/container-manager", "Override container-manager image")
	uiImage := startCmd.String("core-ui", "databoxsystems/core-ui", "Override core ui image")
	arbiterImage := startCmd.String("arbiter", "databoxsystems/arbiter", "Override arbiter image")
	coreNetworkImage := startCmd.String("core-network", "databoxsystems/core-network", "Override container-manager image")
	coreNetworkRelay := startCmd.String("core-network-relay", "databoxsystems/core-network-relay", "Override core-network-relay image")
	appServerImage := startCmd.String("app-server", "databoxsystems/driver-app-store", "Override local manifest driver image")
	exportServerImage := startCmd.String("export-service", "databoxsystems/export-service", "Override export-service image")
	storeImage := startCmd.String("store", "databoxsystems/core-store", "Override core-store image")
	clearSLAdb := startCmd.Bool("flushSLAs", false, "Removes any saved apps or drivers from the SLA database so they will not restart")
	enableLogging := startCmd.Bool("v", false, "Enables verbose logging of the container-manager")
	arch := startCmd.String("arch", "", "Used to override the detected cpu architecture only useful for testing arm64v8 support using docker for mac.")
	sslHostName := startCmd.String("sslHostName", "", "Used to override the detected HostName for use in ssl cert.")
	ReGenerateDataboxCertificates := startCmd.Bool("regenerateCerts", false, "Force databox to regenerate the databox root and certificate")
	stopCmd := flag.NewFlagSet("stop", flag.ExitOnError)
	logsCmd := flag.NewFlagSet("logs", flag.ExitOnError)

	sdkCmd := flag.NewFlagSet("sdk", flag.ExitOnError)
	startSDK := sdkCmd.Bool("start", false, "Use this to start the databox sdk")
	sdkCmdRelease := sdkCmd.String("release", CURRENT_RELEASE, "Databox version to start, can uses tagged versions or latest")
	sdkCmdRegistry := sdkCmd.String("registry", DEFAULT_REGISTRY, "Override the default registry path, where images are pulled form")

	stopSDK := sdkCmd.Bool("stop", false, "Use this to stop the databox sdk")

	testCmd := flag.NewFlagSet("test", flag.ExitOnError)
	testCmdNetwork := testCmd.Bool("network", false, "Perform a network test")

	wipeCmd := flag.NewFlagSet("wipe", flag.ExitOnError)
	wipeCmdRemoveCerts := wipeCmd.Bool("removeCerts", false, "Force databox to remove the databox certificate")
	wipeCmdYes := wipeCmd.Bool("y", false, "Yes I'm sure")

	flag.Parse()

	os.Setenv("DOCKER_API_VERSION", *DOCKER_API_VERSION)
	dockerCli, _ = client.NewEnvClient()

	path = "" //this should not be set outside of a container use --host-path if needed

	if _, err := os.Stat("./certs"); err != nil {
		os.Mkdir("./certs", 0770)
	}
	if _, err := os.Stat("./slaStore"); err != nil {
		os.Mkdir("./slaStore", 0770)
	}

	if len(os.Args) == 1 {
		displayUsage()
		os.Exit(2)
	}

	switch os.Args[1] {
	case "start":
		startCmd.Parse(os.Args[2:])

		libDatabox.Info("Starting Databox " + *startCmdRelease)

		//get some info in the network configuration
		hostname, _ := os.Hostname()
		if *sslHostName != "" {
			hostname = *sslHostName
		}
		ipv4s := removeIPv6addresses(getLocalInterfaceIps())

		cpuArch := ""
		if *arch != "" {
			cpuArch = *arch
		} else if runtime.GOARCH == "amd64" {
			cpuArch = "amd64"
		} else if runtime.GOARCH == "arm64" {
			cpuArch = "arm64v8"
		} else {
			panic("Unsupported CPU architecture ")
		}

		opts := &libDatabox.ContainerManagerOptions{
			Version:               *startCmdRelease,
			SwarmAdvertiseAddress: *startCmdIP,
			ContainerManagerImage: *cmImage + "-" + cpuArch + ":" + *startCmdRelease,
			CoreUIImage:           *uiImage + "-" + cpuArch + ":" + *startCmdRelease,
			ArbiterImage:          *arbiterImage + "-" + cpuArch + ":" + *startCmdRelease,
			CoreNetworkImage:      *coreNetworkImage + "-" + cpuArch + ":" + *startCmdRelease,
			CoreNetworkRelayImage: *coreNetworkRelay + "-" + cpuArch + ":" + *startCmdRelease,
			AppServerImage:        *appServerImage + "-" + cpuArch + ":" + *startCmdRelease,
			ExportServiceImage:    *exportServerImage + "-" + cpuArch + ":" + *startCmdRelease,
			DefaultStoreImage:     *storeImage + "-" + cpuArch + ":" + *startCmdRelease,
			ClearSLAs:             *clearSLAdb,
			DefaultRegistryHost:   *startCmdRegistryHosts,
			DefaultRegistry:       *startCmdRegistry,
			DefaultAppStore:       *appStore,
			EnableDebugLogging:    *enableLogging,
			OverridePasword:       *startCmdPassword,
			InternalIPs:           ipv4s,
			Hostname:              hostname,
			Arch:                  cpuArch,
		}

		externalIP, err := getExternalIP()
		if err != nil {
			libDatabox.Warn("Could not get external IP address, do you have network connectivity? ")
			libDatabox.Warn("Try here for help https://development.robinwinslow.uk/2016/06/23/fix-docker-networking-dns/")
			libDatabox.Warn("Continuing as external connectivity is not always needed ;-)")
			opts.ExternalIP = ipv4s[0]
		} else {
			opts.ExternalIP = externalIP
		}

		if *ReGenerateDataboxCertificates == true {
			libDatabox.Info("Forcing regoration of Databox certificates")
			volume := listDockerVolumesMatching("container-manager-certs")
			removeVolumes(volume)
		}

		Start(opts)
		ShowLogs()
	case "stop":
		libDatabox.Info("Stoping Databox ...")
		stopCmd.Parse(os.Args[2:])
		Stop()
	case "logs":
		logsCmd.Parse(os.Args[2:])
		ShowLogs()
	case "sdk":
		sdkCmd.Parse(os.Args[2:])

		if *startSDK == true {
			sdkCmd.Parse(os.Args[2:])
			libDatabox.Info("Starting Databox SDK")
			StartSDK(*sdkCmdRegistry, *sdkCmdRelease)

		} else if *stopSDK == true {
			libDatabox.Info("Stoping Databox SDK")
			StopSDK()
		} else {
			sdkCmd.Usage()
		}
	case "wipe":
		wipeCmd.Parse(os.Args[2:])

		if *wipeCmdYes == false {
			fmt.Println("No changes made you did not add the -y flag")
			return
		}

		fmt.Println("This will wipe all databox data. Are you sure?")
		fmt.Println("you have 10 seconds to change your mind (ctrl+c to exit)")
		fmt.Println("\n If you do not exit I will delete: \n")
		volumes := listDockerVolumesMatching("-core-store")
		for _, v := range volumes {
			fmt.Println(v.Name)
		}
		time.Sleep(10 * time.Second)

		fmt.Println("Stoping Databox ...")
		Stop()

		fmt.Println("Removing all app and driver data ....")

		removeVolumes(volumes)

		if *wipeCmdRemoveCerts {
			fmt.Println("Removing certificates ....")
			volume := listDockerVolumesMatching("container-manager-certs")
			removeVolumes(volume)
		}

	case "test":

		testCmd.Parse(os.Args[2:])

		if *testCmdNetwork {

			fmt.Println("Listing addresses")
			fmt.Println(removeIPv6addresses(getLocalInterfaceIps()))
			fmt.Println("Testing DNS")

			adr, err := net.LookupHost("google.com")
			if err != nil {
				libDatabox.Err("Could not look up google.com is DNS broken? " + err.Error())
			} else {
				fmt.Println("google.com's ips are ", adr)
			}
			fmt.Println("Trying to get external IP ")
			externalIP, err := getExternalIP()
			if err != nil {
				libDatabox.Err("Could not get external IP address, do you have network connectivity? " + err.Error())
			} else {
				fmt.Println("external IP is ", externalIP)
			}
		} else {
			testCmd.Usage()
		}
	default:
		displayUsage()
		os.Exit(2)
	}

}

func displayUsage() {
	fmt.Println(`
		databox [cmd]
		Usage:
			start - start databox
			stop - stop databox
			logs - view databox logs
			sdk  - manage the databox sdk
			wipe  - remove databox data

		Use databox [cmd] help to see more options
		`)
}

func Start(opt *libDatabox.ContainerManagerOptions) {

	_, err := dockerCli.SwarmInit(context.Background(), swarm.InitRequest{
		ListenAddr:    "127.0.0.1",
		AdvertiseAddr: opt.SwarmAdvertiseAddress,
	})
	libDatabox.ChkErrFatal(err)

	os.Remove("/tmp/databox_relay")
	err = syscall.Mkfifo("/tmp/databox_relay", 0666)
	libDatabox.ChkErrFatal(err)

	createContainerManager(opt)

}

func Stop() {

	_, err := dockerCli.SwarmInspect(context.Background())
	if err != nil {
		//Not in swarm mode databox is not running
		return
	}
	filters := filters.NewArgs()
	filters.Add("label", "databox.type")

	services, err := dockerCli.ServiceList(context.Background(), types.ServiceListOptions{Filters: filters})
	libDatabox.ChkErr(err)

	if len(services) > 0 {
		for _, service := range services {
			libDatabox.Info("Removing old databox service " + service.Spec.Name)
			err := dockerCli.ServiceRemove(context.Background(), service.ID)
			libDatabox.ChkErr(err)
		}
	}

	dockerCli.SwarmLeave(context.Background(), true)

	containers, err := dockerCli.ContainerList(context.Background(), types.ContainerListOptions{Filters: filters})
	libDatabox.ChkErr(err)

	if len(containers) > 0 {
		for _, container := range containers {
			libDatabox.Info("Removing old databox container " + container.Image)
			err := dockerCli.ContainerStop(context.Background(), container.ID, nil)
			libDatabox.ChkErr(err)
			err = dockerCli.ContainerRemove(context.Background(), containers[0].ID, types.ContainerRemoveOptions{Force: true})
			libDatabox.ChkErr(err)
		}
	}

}

func listDockerVolumesMatching(filterString string) []*types.Volume {
	f := filters.NewArgs()
	f.Add("name", filterString)
	volumes, err := dockerCli.VolumeList(context.Background(), f)
	libDatabox.ChkErrFatal(err)

	return volumes.Volumes
}

func removeVolumes(volumes []*types.Volume) {
	for _, v := range volumes {
		fmt.Print("Removing ", v.Name, "....")
		err := dockerCli.VolumeRemove(context.Background(), v.Name, true)
		libDatabox.ChkErrFatal(err)
		fmt.Println("   Done! ")
	}
}

func createContainerManager(options *libDatabox.ContainerManagerOptions) {

	portConfig := []swarm.PortConfig{
		swarm.PortConfig{
			TargetPort:    443,
			PublishedPort: 443,
			PublishMode:   "host",
		},
		swarm.PortConfig{
			TargetPort:    80,
			PublishedPort: 80,
			PublishMode:   "host",
		},
	}

	//create options secret
	optionsJSON, err := json.Marshal(options)
	libDatabox.ChkErrFatal(err)
	secretCreateResponse, err := dockerCli.SecretCreate(context.Background(), swarm.SecretSpec{
		Annotations: swarm.Annotations{
			Name: "DATABOX_CM_OPTIONS",
		},
		Data: optionsJSON,
	})
	libDatabox.ChkErrFatal(err)

	cmOptionsSecret := swarm.SecretReference{
		SecretID:   secretCreateResponse.ID,
		SecretName: "DATABOX_CM_OPTIONS",
		File: &swarm.SecretReferenceFileTarget{
			Name: "DATABOX_CM_OPTIONS",
			UID:  "0",
			GID:  "0",
			Mode: 0444,
		},
	}

	service := swarm.ServiceSpec{
		TaskTemplate: swarm.TaskSpec{
			ContainerSpec: &swarm.ContainerSpec{
				Image:    options.ContainerManagerImage,
				Hostname: "container-manager",
				Labels:   map[string]string{"databox.type": "container-manager"},
				Env: []string{
					"DATABOX_ARBITER_ENDPOINT=tcp://arbiter:4444",
					"DATABOX_SDK=0",
				},
				Mounts: []mount.Mount{
					mount.Mount{
						Type:   mount.TypeBind,
						Source: "/var/run/docker.sock",
						Target: "/var/run/docker.sock",
					},
					mount.Mount{
						Source: "container-manager-certs",
						Target: "/certs",
						Type:   mount.TypeVolume,
					},
				},
				Secrets: []*swarm.SecretReference{&cmOptionsSecret},
			},
			Placement: &swarm.Placement{
				Constraints: []string{"node.role == manager"},
			},
		},
		EndpointSpec: &swarm.EndpointSpec{
			Mode:  "dnsrr",
			Ports: portConfig,
		},
	}

	service.Name = "container-manager"

	serviceOptions := types.ServiceCreateOptions{}

	pullImage(service.TaskTemplate.ContainerSpec.Image, options)
	_, err = dockerCli.ServiceCreate(context.Background(), service, serviceOptions)
	libDatabox.ChkErr(err)

}

func pullImage(image string, options *libDatabox.ContainerManagerOptions) {

	needToPull := true

	//do we have the image on disk?
	images, _ := dockerCli.ImageList(context.Background(), types.ImageListOptions{})
	for _, i := range images {
		for _, tag := range i.RepoTags {
			if image == tag {
				//we have the image no need to pull it !!
				needToPull = false
				break
			}
		}
	}

	//is it from the default registry (databoxsystems or whatever we overroad with) and tagged with latest?
	if strings.Contains(image, options.DefaultRegistry) == true && strings.Contains(image, ":latest") == true {
		//its in the default registry and has the :latest tag lets pull it to make sure we are up-to-date
		needToPull = true
	}

	if needToPull == true {
		libDatabox.Info("Pulling Image " + image)
		reader, err := dockerCli.ImagePull(context.Background(), image, types.ImagePullOptions{})
		libDatabox.ChkErr(err)
		io.Copy(ioutil.Discard, reader)
		libDatabox.Info("Done pulling Image " + image)
		reader.Close()
	}
}

func removeContainer(name string) {
	filters := filters.NewArgs()
	filters.Add("name", name)
	containers, clerr := dockerCli.ContainerList(context.Background(), types.ContainerListOptions{
		Filters: filters,
		All:     true,
	})
	libDatabox.ChkErr(clerr)

	if len(containers) > 0 {
		rerr := dockerCli.ContainerRemove(context.Background(), containers[0].ID, types.ContainerRemoveOptions{Force: true})
		libDatabox.ChkErr(rerr)
	}
}

func getLocalInterfaceIps() []string {
	IPs := []string{}
	ifaces, _ := net.Interfaces()
	for _, i := range ifaces {
		addrs, _ := i.Addrs()
		for _, addr := range addrs {
			switch v := addr.(type) {
			case *net.IPNet:
				IPs = append(IPs, v.IP.String())
			case *net.IPAddr:
				IPs = append(IPs, v.IP.String())
			}
		}
	}
	return IPs
}

func removeIPv6addresses(addresses []string) []string {
	var filteredIPs []string
	for _, ip := range addresses {
		parsedIP := net.ParseIP(ip)
		if parsedIP.To4() != nil {
			filteredIPs = append(filteredIPs, ip)
		}
	}
	return filteredIPs
}

func getExternalIP() (string, error) {
	var netTransport = &http.Transport{
		Dial: (&net.Dialer{
			Timeout: 5 * time.Second,
		}).Dial,
		TLSHandshakeTimeout: 5 * time.Second,
	}
	var netClient = &http.Client{
		Timeout:   time.Second * 10,
		Transport: netTransport,
	}
	response, err := netClient.Get("http://whatismyip.akamai.com/")
	if err != nil {
		libDatabox.Debug("[ERROR] " + err.Error())
		return "", err
	}
	ip, err := ioutil.ReadAll(response.Body)
	if err != nil {
		libDatabox.Debug("[ERROR] " + err.Error())
		return "", err
	}
	response.Body.Close()
	libDatabox.Debug("External IP found " + string(ip))
	return string(ip), nil
}
